lib/ipam/ipam_block_reader_writer.go

```diff
@@ -347,6 +345,76 @@ func (rw blockReaderWriter) releaseBlockAffinity(ctx context.Context, host strin
 	return nil
 }
 
+func (rw blockReaderWriter) queryAffinity(ctx context.Context, k model.Key, revision string) (*model.KVPair, error) {
+	aff, err := rw.client.Get(ctx, k, revision)
+	if err != nil {
+		return nil, err
+	}
+
+	if aff.Value.(*model.BlockAffinity).State != model.StateDeleted {
+		// We're all good - return the afifnity.
+		return aff, nil
+	}
+
+	// If the affinity is marked as deleted, we need to delete it.
+	if err := rw.deleteAffinity(ctx, aff); err != nil {
```

Comment:
Doesn't it seem a little precarious to be deleting a resource on a query operation?

Reply:
Wouldn't it be better just to return a `DoesNotExist` error and trust that whomever marked the affinity for deletion will follow through with it?

Reply:
I'm not sure what you mean by precarious in this case :D 

We actually can't rely on this being deleted by the code that marked it for deletion - it's totally possible that the other process successfully marked it as deleted, but then failed to delete the resource for some reason, meaning we need to clean up after it the next time this data is accessed.